paddle/fluid/operators/fake_quantize_op.h

```diff
@@ -0,0 +1,163 @@
+    for (int i = 0; i < n; ++i) {
+      if (in[i] > max) {
+        out[i] = max;
+        ++num_saturate;
+      } else if (in[i] < min) {
+        out[i] = min;
+        ++num_saturate;
+      } else {
+        out[i] = in[i];
+      }
+    }
+    return num_saturate;
+  }
+
+  virtual void Compute(const framework::ExecutionContext& context) const {
+    auto* tensor = context.Output<framework::Tensor>("Out");
+    auto* in = context.Input<framework::Tensor>("X");
+    const bool is_test = context.Attr<bool>("is_test");
+    tensor->mutable_data<T>(in->place());
+
+    auto* oms_tensor = context.Output<framework::Tensor>("OutMovingScale");
+    oms_tensor->mutable_data<T>(in->place());
+
+    auto quantize_type =
+        static_cast<std::string>(context.Attr<std::string>("quantize_type"));
+    if (quantize_type == std::string("range_abs_max")) {
+      auto* oss_tensor = context.Output<framework::Tensor>("OutScales");
+      oss_tensor->mutable_data<T>(
+          context.Input<framework::Tensor>("InScales")->place());
+      auto* oci_tensor = context.Output<framework::Tensor>("OutCurrentIter");
+      oci_tensor->mutable_data<T>(
+          context.Input<framework::Tensor>("InCurrentIter")->place());
+    }
+
+    T scale = T(1);
+    int window_size = context.Attr<int>("window_size");
+    int bit_length = context.Attr<int>("bit_length");
+    int bin_cnt = std::pow(2, bit_length - 1) - 1;
+    LOG(ERROR) << "bin_cnt:" << bin_cnt;
+
+    if (quantize_type == std::string("abs_max")) {
+      auto* saving_scale = context.Output<framework::Tensor>("OutMovingScale");
+      scale = T(find_abs_max(const_cast<framework::Tensor*>(in), in->numel()));
+      saving_scale->mutable_data<T>(platform::CPUPlace())[0] = scale;
+
+      auto& device_ctx = context.template device_context<DeviceContext>();
+      auto* scale_list = context.Output<framework::Tensor>("OutScales");
+      math::SetConstant<DeviceContext, T> scalar;
+      scale_list->mutable_data<T>(context.GetPlace());
+      scalar(device_ctx, scale_list, static_cast<T>(0));
+      auto* iter = context.Output<framework::Tensor>("OutCurrentIter");
+      iter->mutable_data<T>(context.GetPlace());
+      scalar(device_ctx, iter, static_cast<T>(0));
+    } else if (quantize_type == std::string("range_abs_max")) {
+      auto* moving_scale = context.Input<framework::Tensor>("InMovingScale");
+      if (is_test) {
+        scale = moving_scale->data<T>()[0];
+      } else {
+        auto* it = context.Input<framework::Tensor>("InCurrentIter");
+        auto* iter = context.Output<framework::Tensor>("OutCurrentIter");
+        const int* last_iter = it->data<int>();
+        int* current_iter = iter->mutable_data<int>(platform::CPUPlace());
+        auto* scale_list = context.Output<framework::Tensor>("OutScales");
+        auto* saving_scale =
+            context.Output<framework::Tensor>("OutMovingScale");
+        scale = find_abs_max(const_cast<framework::Tensor*>(in), in->numel());
```

Comment:
Here can use Eigen's method：

abs: https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-abs
max: https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-maximum

Reply:
this cwiseMax is an elemwise max operation, i need a reduce max op .

Reply:
Please refer https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-reduceconst-dimensions-new_dims-const-reducer-reducer

or 

 https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-maximum

Reply:
done